src/Features/LanguageServer/ProtocolUnitTests/Diagnostics/PullDiagnosticTests.cs

```diff
@@ -0,0 +1,469 @@
+                workspace, previousResults: CreateDiagnosticParamsFromPreviousReports(results));
+
+            Assert.Equal(2, results2.Length);
+            Assert.Empty(results2[0].Diagnostics);
+            Assert.Null(results2[1].Diagnostics);
+
+            Assert.NotEqual(results[0].ResultId, results2[0].ResultId);
+            Assert.Equal(results[1].ResultId, results2[1].ResultId);
+        }
+
+        [Fact]
+        public async Task TestWorkspaceDiagnosticsRemainAfterErrorIsNotFixed()
+        {
+            var markup1 =
+@"class A {";
+            var markup2 = "";
+            using var workspace = CreateTestWorkspaceWithDiagnostics(
+                 new[] { markup1, markup2 }, BackgroundAnalysisScope.FullSolution);
+
+            var results = await RunGetWorkspacePullDiagnosticsAsync(workspace);
+
+            Assert.Equal(2, results.Length);
+            Assert.Equal("CS1513", results[0].Diagnostics.Single().Code);
+            Assert.Equal(new Position { Line = 0, Character = 9 }, results[0].Diagnostics.Single().Range.Start);
+
+            Assert.Empty(results[1].Diagnostics);
+
+            var buffer = workspace.Documents.First().GetTextBuffer();
+            buffer.Insert(0, " ");
+
+            var results2 = await RunGetWorkspacePullDiagnosticsAsync(workspace);
+
+            Assert.Equal("CS1513", results2[0].Diagnostics.Single().Code);
+            Assert.Equal(new Position { Line = 0, Character = 10 }, results2[0].Diagnostics.Single().Range.Start);
+
+            Assert.Empty(results2[1].Diagnostics);
+            Assert.NotEqual(results[1].ResultId, results2[1].ResultId);
+        }
+
+        [Fact]
+        public async Task TestStreamingWorkspaceDiagnostics()
+        {
+            var markup1 =
+@"class A {";
+            var markup2 = "";
+            using var workspace = CreateTestWorkspaceWithDiagnostics(
+                 new[] { markup1, markup2 }, BackgroundAnalysisScope.FullSolution);
+
+            var results = await RunGetWorkspacePullDiagnosticsAsync(workspace);
+
+            Assert.Equal(2, results.Length);
+            Assert.Equal("CS1513", results[0].Diagnostics.Single().Code);
+            Assert.Equal(new Position { Line = 0, Character = 9 }, results[0].Diagnostics.Single().Range.Start);
+
+            var progress = BufferedProgress.Create<DiagnosticReport[]>(null);
+            results = await RunGetWorkspacePullDiagnosticsAsync(workspace, progress: progress);
+
+            Assert.Null(results);
+            Assert.Equal("CS1513", progress.GetValues()![0][0].Diagnostics![0].Code);
+        }
+
+        #endregion
+
+        private static async Task<DiagnosticReport[]> RunGetDocumentPullDiagnosticsAsync(
+            TestWorkspace workspace, Document document,
+            string? previousResultId = null,
+            IProgress<DiagnosticReport[]>? progress = null)
+        {
+            var solution = document.Project.Solution;
+            var queue = CreateRequestQueue(solution);
```

Comment:
For tests that call RunGetDocumentPullDiagnosticsAsync multiple times, should we use a single queue instead of creating a new one each time?

Reply:
yup, i can do that.